content/doc/book/pipeline/overview.adoc

```diff
@@ -1,4 +1,297 @@
+  stages {
+    stage('Build') {
+      steps {
+        sh 'npm --version'
+        sh 'npm install'
+        sh 'npm test'
+      }
+    }
+  }
+}
+// Script //
+node {
+  stage('Build') {
+    sh 'npm --version'
+    sh 'npm install'
+    sh 'npm test'
+  }
+}
+----
+
+
+*Ruby*
+
+[pipeline]
+----
+// Declarative //
+pipeline {
+  agent { label '' }
+  stages {
+    stage('Build') {
+      steps {
+        sh 'ruby --version'
+        sh 'bundle install'
+      }
+    }
+  }
+}
+// Script //
+node {
+  stage('Build') {
+    sh 'ruby --version'
+    sh 'bundle install'
+  }
+}
+----
+
+*Python*
+
+[pipeline]
+----
+// Declarative //
+pipeline {
+  agent { label ' }
+  stages {
+    stage('Build') {
+      steps {
+        sh 'pip --version'
+        sh 'python --version'
+      }
+    }
+  }
+}
+// Script //
+node {
+  stage('Build') {
+    sh 'pip --version'
+    sh 'python --version'
+  }
+}
+----
+
+*PHP*
+
+[pipeline]
+----
+// Declarative //
+pipeline {
+  agent { label '' }
+  stages {
+    stage('Build') {
+      steps {
+        sh 'php --version'
+      }
+    }
+  }
+}
+// Script //
+node {
+  stage('Build') {
+    sh 'php --version'
+  }
+}
+----
+
+[[agents]]
+== Working with Agents
+
+In the simple examples above, you will have seen the `agent` directive in use with
+Declarative Pipeline and `node` with Scripted Pipeline. `agent` and `node` perform
+similar functions in each respective syntax to define where the Pipeline is executed,
+but `agent` includes additional inputs to easily work with Docker.
+As a distributed system most _work_ a Pipeline performs is done in the context
+of one or more nodes. Defining a `node` or `agent` does two things:
+
+* Schedules the steps contained within the block to run by adding an item to the Jenkins queue.
+As soon as an executor is free on a node, the steps will run.
+* Creates a workspace (a directory specific to that particular Pipeline) where
+work can be done on files checked out from source control.
+
+There are several ways to define agents to use in your Pipeline.
+
+=== Using Labels
+This is useful for when you want to specify a particular machine size, or requirement. Use `agent { label 'windows' }` to specify that you want your build to run on an agent that has been labelled as `windows`. If you make this `label ''` then it will run on any agent connected to your Jenkins master that is available.
+
+Example (will use the `master` node):
+[pipeline]
+----
+// Declarative //
+pipeline {
+    agent {
+        label 'master'
+    }
+    stages {
+        stage('testing 123') {
+            steps {
+                sh 'echo hello from master node'
+            }
+        }
+    }
+}
+// Script //
+node('master') {
+  stage('testing 123') {
+    sh 'echo hello from master node'
+  }
+}
+----
+
+=== Using Docker
+Declarative Pipeline is designed to easily use Docker images and containers to run
+inside. This allows the Pipeline to define the build and test environment and tools
```

Comment:
inside what? Do you mean that Declarative Pipeline can run inside Docker images and containers?